tools/generate_release_notes.py

```diff
@@ -1,246 +1,498 @@
+                )
+        all_pull_requests.update(commit_pull_requests)
+    return all_pull_requests
+
+
+@dataclass(frozen=True, eq=True)
+class CoAuthor:
+    """Represents a co-author for which the GitHub login is not known.
+
+    Hashing and comparing only takes into account the `name` attribute.
+    """
 
-for tag in repository.get_tags():
-    if tag.name == args.from_commit:
-        previous_tag = tag
-        break
-else:
-    raise RuntimeError(f"Desired tag ({args.from_commit}) not found")
+    name: str
+    email: str
+    commit: Commit = field(compare=False)
 
-# For some reason, go get the github commit from the commit to get
-# the correct date
-github_commit = previous_tag.commit.commit
-previous_tag_date = datetime.strptime(
-    github_commit.last_modified, "%a, %d %b %Y %H:%M:%S %Z"
-)
 
+def find_coauthors(commit: Commit) -> set[CoAuthor]:
+    """Find co-author in a commit message.
 
-all_commits = list(
-    tqdm(
-        repository.get_commits(sha=args.to_commit, since=previous_tag_date),
-        desc=f"Getting all commits between {args.from_commit} " f"and {args.to_commit}",
+    These are matched by looking for occurrences of "Co-authored-by:" bollowed by a name
+    and email address in a commit message. Unfortunately, GitHub's API doesn't provide
+    a way yet to extract co-authors as NamedUsers from a commit.
+    See also  :func:`try_replace_coauthors`
+    """
+    co_author_regex = re.compile(
+        r"Co-authored-by: (?P<name>[^<]+) <(?P<email>[^>]+)>", flags=re.MULTILINE
     )
-)
-all_hashes = {c.sha for c in all_commits}
-
-authors = set()
-reviewers = set()
-committers = set()
-users = dict()  # keep track of known usernames
+    message = commit.commit.message
+    matches = list(co_author_regex.finditer(message))
+    expected_count = message.count("Co-authored-by:")
+    if len(matches) != expected_count:
+        logger.warning(
+            "expected %i co-authors, regex found %i in %s",
+            expected_count,
+            len(matches),
+            commit.html_url,
+        )
+    coauthors = {
+        CoAuthor(name=m["name"], email=m["email"], commit=commit) for m in matches
+    }
+    return coauthors
+
+
+def contributors(
+    commits: Iterable[Commit], pull_requests: Iterable[PullRequest]
+) -> tuple[set[NamedUser], set[CoAuthor], set[NamedUser]]:
+    """Fetch commit authors, co-authors and reviewers.
+
+    `authors` are users which created a commit.
+    `coauthors` are name, and email pairs that signify the co-author of a commit, see
+    :func:`find_coauthors` and :func:`try_replace_coauthors`.
+    `reviewers` are users, who added reviews to a merged pull request or created the
+    merge commit for one.
+    """
+    authors = set()
+    reviewers = set()
+    coauthors = set()
 
+    for commit in commits:
+        if commit.author:
+            authors.add(commit.author)
+        if commit.committer:
+            reviewers.add(commit.committer)
```

Comment:
Note, I am sorting committers as reviewers here. The "logic" is that that action feels more administrative and may indicate a maintainer directly merging after an implicit but not public review. Previously, committers where ignored I think.